CORE-465: Add support for signatures in raw format

ewerx opened the pull request:
- add interface for DER and and Raw signatures in BRCryptoSigner
- add BRKeySignRaw and DER to Raw/JOSE conversion for signatures in BRKey

EBGToo commented:
As this impacts `support/BRKey.*`, there needs to be a test case in `bitcoin/test.c` in `BRKeyTests()` near L1280.  The test there can be identical to the test in BRCryptoCommonTests.swift.

michael-brd commented on support/BRKey.c:
```diff
@@ -364,7 +372,84 @@ size_t BRKeyLegacyAddr(BRKey *key, char *addr, size_t addrLen, BRAddressParams p
 
-// signs md with key and writes signature to sig
+// convert signature from DER to raw encoding
+int BRSignatureDERToRaw (const uint8_t *derSig, size_t derSigLen, uint8_t *rawSig, size_t rawSigLen)
```
I think this function could be static, what do you think?

michael-brd commented on support/BRKey.c:
```diff
@@ -364,7 +372,84 @@ size_t BRKeyLegacyAddr(BRKey *key, char *addr, size_t addrLen, BRAddressParams p
+    
+    //Convert to unsigned. Should match DER length - offset
+    int encodedLength = derSig[offset++] & 0xff;
+    assert(encodedLength == derSigLen - offset);
```
I think we want to fail out (maybe in addition to an assert) since the assert may be a NOP depending on how this source is built.

michael-brd commented on support/BRKey.c:
```diff
@@ -364,7 +372,84 @@ size_t BRKeyLegacyAddr(BRKey *key, char *addr, size_t addrLen, BRAddressParams p
+// convert signature from DER to raw encoding
+int BRSignatureDERToRaw (const uint8_t *derSig, size_t derSigLen, uint8_t *rawSig, size_t rawSigLen)
+{
+    const int sigLen = 64;
```
I think we should avoid using signed types throughout here. I know the original Java used signed types, but unsigned types aren't readily available in Java. By sticking to unsigned types and verifying every bit of arithmetic for overflows, I think we will be better off.

ewerx commented:
After discussion with @voisine, I renamed the new functions `BRKeySignJOSE` and `BRKeyVerifyJOSE` and refactored them to use the existing `secp256k1_ecdsa_signature_serialize_compact` serialization function instead of the DER to JOSE conversion.
Added tests.
Updated CryptoSwift interface and test.

EBGToo commented on support/BRKey.c:
```diff
@@ -53,6 +53,14 @@
 #pragma clang diagnostic pop
 #pragma GCC diagnostic pop
 
+#if !defined (MAX)
```
Are MAX and MIN needed now?

ewerx replied:
nope, removed.

EBGToo commented on Swift/BRCrypto/common/BRCryptoSigner.swift:
```diff
@@ -77,6 +78,15 @@ public enum CoreSigner: Signer {
+                
+            case .basicJOSE:
+                let targetCount = BRKeySignJOSE (cryptoKeyGetCore(key), nil, 0, digestUInt256)
+                if 0 == targetCount { /* error */ }
```
Do you have a sense if these `0 == targetCount` errors are possible?  If so, should the return be `Data?` (or Result<Data,SignerError>` (if `SignerError` is enumerable)?

ewerx replied:
I stepped through all the signing/serialization functions and they will always return 1. The only scenario where `BRKeySign` returns 0 is if the length passed in less than required and the target buffer is non-nil. I replaced the `targetCount == 0` error checks with asserts.
